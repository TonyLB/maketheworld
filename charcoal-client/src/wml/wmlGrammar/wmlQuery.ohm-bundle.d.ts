// AUTOGENERATED FILE
// This file was generated from wmlQuery.ohm by `ohm generateBundles`.

import {
  ActionDict,
  Grammar,
  IterationNode,
  Node,
  NonterminalNode,
  Semantics,
  TerminalNode
} from 'ohm-js';

export interface WMLQuerySelectorActionDict<T> extends ActionDict<T> {
  WMLQueryContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  WMLLegalTag?: (this: NonterminalNode, arg0: TerminalNode) => T;
  TagAncestry?: (this: NonterminalNode, arg0: IterationNode) => T;
}

export interface WMLQuerySelectorSemantics extends Semantics {
  addOperation<T>(name: string, actionDict: WMLQuerySelectorActionDict<T>): this;
  extendOperation<T>(name: string, actionDict: WMLQuerySelectorActionDict<T>): this;
  addAttribute<T>(name: string, actionDict: WMLQuerySelectorActionDict<T>): this;
  extendAttribute<T>(name: string, actionDict: WMLQuerySelectorActionDict<T>): this;
}

export interface WMLQuerySelectorGrammar extends Grammar {
  createSemantics(): WMLQuerySelectorSemantics;
  extendSemantics(superSemantics: WMLQuerySelectorSemantics): WMLQuerySelectorSemantics;
}

declare const grammar: WMLQuerySelectorGrammar;
export default grammar;

