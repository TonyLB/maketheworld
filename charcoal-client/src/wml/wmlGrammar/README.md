The wml.ohm-bundle files are generated:  Only the wml.ohm file is original.

To regenerate (e.g. after changing the wml.ohm file), navigate to the src/wml directory, and
execute this command:  npx ohm generateBundles --withTypes 'wmlGrammar/*.ohm'