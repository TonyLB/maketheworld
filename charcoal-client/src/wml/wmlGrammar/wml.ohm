WorldMarkupLangage {

    //
    // This section defines the tags that make up the WML language
    //

    WMLFileContents = AssetExpression
            | StoryExpression
            | CharacterExpression

    CharacterExpression = TagExpression<"Character", "key", ("player" | "fileName" | "zone" | "subFolder"), none, none, CharacterLegalContents, none>
    CharacterLegalContents = ImageExpression
            | LiteralNameExpression
            | PronounsExpression
            | LiteralValueTag<"OneCoolThing">
            | LiteralValueTag<"Outfit">
            | LiteralValueTag<"FirstImpression">

    PronounsExpression = TagExpression<"Pronouns", none, ("subject" | "object" | "possessive" | "adjective" | "reflexive"), none, none, none, none>
            | TagSelfClosing<"Pronouns", none, ("subject" | "object" | "possessive" | "adjective" | "reflexive"), none, none, none>

    StoryExpression = TagExpression<"Story", "key", ("fileName" | "zone" | "subFolder" | "player"), none, "instance", AssetLegalContents, none>

    AssetExpression = TagExpression<"Asset", "key", ("fileName" | "zone" | "subFolder" | "player"), none, none, AssetLegalContents, none>
    AssetLegalContents = ImportExpression
            | VariableExpression
            | ComputedExpression
            | ActionExpression
            | LayerExpression
            | ConditionExpression
            | MapExpression
            | RoomExpression
            | ExitExpression
            | FeatureExpression
            | LiteralNameExpression
            | string

    MapExpression = TagExpression<"Map", "key", none, none, none, MapLegalContents, none>
    MapLegalContents = MapRoomExpression
            | NameExpression
            | ImageExpression

    ImageExpression = TagExpression<"Image", "key", "fileURL", none, none, none, none>
            | TagSelfClosing<"Image", "key", "fileURL", none, none, none>

    ImportExpression = TagExpression<"Import", "from", none, none, none, ImportLegalContents, none>
    ImportLegalContents = UseExpression

    UseExpression = TagExpression<"Use", ("key" | "as"), "type", none, none, none, none>
            | TagSelfClosing<"Use", ("key" | "as"), "type", none, none, none>

    DependencyExpression = TagExpression<"Depend", "on", none, none, none, none, none>
            | TagSelfClosing<"Depend", "on", none, none, none, none>

    VariableExpression = TagExpression<"Variable", "key", none, "default", none, none, none>
            | TagSelfClosing<"Variable", "key", none, "default", none, none>

    ComputedExpression = TagExpression<"Computed", "key", none, "src", none, DependencyExpression, none>
            | TagSelfClosing<"Computed", "key", none, "src", none, none>

    ActionExpression = TagExpression<"Action", "key", none, "src", none, none, none>
            | TagSelfClosing<"Action", "key", none, "src", none, none>

    ConditionExpression = TagExpression<"Condition", none, none, "if", none, ConditionLegalContents, none>
    ConditionLegalContents = RoomExpression
            | MapExpression
            | ConditionExpression
            | TextContents
            | LayerExpression
            | ExitExpression
            | FeatureExpression
            | DependencyExpression

    LayerExpression = TagExpression<"Layer", "key", none, none, none, LayerLegalContents, none>
    LayerLegalContents = RoomExpression
            | ConditionExpression
            | ExitExpression

    ExitExpression = TagExpression<"Exit", ("key" | "to" | "from"), none, none, none, string, none>
            | TagSelfClosing<"Exit", ("key" | "to" | "from"), none, none, none, none>

    LinkExpression = TagExpression<"Link", "to", none, none, none, string, spaceCompressor>

    MapRoomExpression = TagExpression<"Room", "key", ("display" | "x" | "y"), none, "global", RoomContents, none>
            | TagSelfClosing<"Room", "key", ("display" | "x" | "y"), none, "global", none>
    RoomExpression = TagExpression<"Room", "key", none, none, "global", RoomContents, none>
            | TagSelfClosing<"Room", "key", "display", none, "global", none>
    RoomContents = DescriptionExpression
            | NameExpression
            | ExitExpression
            | FeatureExpression

    FeatureExpression = TagExpression<"Feature", "key", none, none, "global", FeatureContents, none>
            | TagSelfClosing<"Feature", "key", "display", none, "global", none>
    FeatureContents = DescriptionExpression
            | NameExpression

    DescriptionExpression = TagExpression<"Description", none, "display", none, ("spaceBefore" | "spaceAfter"), DescriptionContents, none>
    DescriptionContents = DescriptionTextContents
            | LinkExpression
            | LineBreakExpression

    LineBreakExpression = "<br" "/>"

    LiteralNameExpression = LiteralValueTag<"Name">
    NameExpression = TagExpression<"Name", none, none, none, ("spaceBefore" | "spaceAfter"), TextContents, none>

    LiteralValueTag<label> = TagExpression<label, none, none, none, none, string, none>

    ValueTag<label> = TagExpression<label, none, none, none, none, TextContents, none>
    DescriptionTextContents = string
    TextContents = string | embeddedJSExpression

    //
    // The following define the ways in which tags can be structured
    //

    TagExpression<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs, LegalContents, endSpace> = TagOpen<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> LegalContents* tagClose<tag> #(endSpace)?
    string = (stringText | spaceCompressor)+
    stringText = (~("<" | "{" | space) any | "\\<" | "\\{")+
    legalKey = letter (letter | digit | "_")+
    spaceCompressor = (space)+
    TagSelfClosing<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs, endSpace> = "<" tag TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs>* "/>" #(endSpace)?
    TagOpen<tag, tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> = "<" tag TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs>* ">"
    TagArgument<tagKeyArgs, tagLiteralArgs, tagProgramArgs, tagBooleanArgs> = TagArgumentKey<tagKeyArgs> | TagLiteralArgument<tagLiteralArgs> | TagProgramArgument<tagProgramArgs> | tagBooleanArgument<tagBooleanArgs>
    TagLiteralArgument<tagLiteralArgs> = tagArgumentQuoted<tagLiteralArgs>
    TagProgramArgument<tagProgramArgs> = tagArgumentQuoted<tagProgramArgs> | tagArgumentBracketed<tagProgramArgs>
    tagBooleanArgument<tagBooleanArgs> = tagBooleanArgs &(space | ">" | "/>")
    tagArgumentQuoted<tagLegalArgs> = tagLegalArgs "=\"" tagArgValueQuoted
    TagArgumentKey<tagLegalArgs> = #(tagLegalArgs "=(") legalKey ")"
    tagArgumentBracketed<tagLegalArgs> = tagLegalArgs "={" jsExpression* "}"
    tagArgValueQuoted = ("\\\"" | ~"\"" any)* "\""
    tagClose<tag> = "</" tag ">"
    none = ~">" ">"

    //
    // embeddedJSExpression parses from the starting open bracket through to the
    //   *MATCHING* close bracket, using a subset of JS syntax that concentrates
    //   on all the ways that JS could let you mess up that matching (quotes,
    //   nesting, and template literals figuring high on the list)
    //
    // While we could, technically, validate the entire ES5 JS Syntax (after all,
    // Ohm publishes their schema), it wouldn't gain us any value here, and it
    // has a quite noticeable resource cost:  Compiling that schema and executing
    // it slows the perception Lambda function down to a crawl.  If a need for
    // such pre-processing emerges, look into how to optimize compiling and executing
    // very large Ohm schemata.
    //
    embeddedJSExpression = "{" jsExpression* "}"
    jsExpression = embeddedJSExpression
                    | jsDoubleQuotedString
                    | jsSingleQuotedString
                    | jsTemplateString
                    | jsText
    jsDoubleQuotedString = "\"" jsNoDoubleQuote* "\""
    jsNoDoubleQuote = "\\\"" | (~"\"" any)
    jsSingleQuotedString = "\'" jsNoSingleQuote* "\'"
    jsNoSingleQuote = "\\\'" | (~"\'" any)
    jsTemplateString = "`" (jsNoBackQuote | jsTemplateEvaluation )* "`"
    jsTemplateEvaluation = "${" jsExpression "}"
    jsNoBackQuote = "\\`" | (~("`" | "${") any)
    jsText = ("\\{" | "\\\"" | "\\'" | "\\`" | "\\}" | (~("{" | "\"" | "\'" | "`" | "}") any))+

    // Override Ohm's built-in definition of space.  Cribbed from es5.ohm
    space := whitespace | lineTerminator | comment

	sourceCharacter = any
    whitespace = "\t"
                | "\x0B"    -- verticalTab
                | "\x0C"    -- formFeed
                | " "
                | "\u00A0"  -- noBreakSpace
                | "\uFEFF"  -- byteOrderMark
                | unicodeSpaceSeparator

	unicodeSpaceSeparator = "\u2000".."\u200B" | "\u3000"

    lineTerminator = "\n" | "\r" | "\u2028" | "\u2029"
    lineTerminatorSequence = "\n" | "\r" ~"\n" | "\u2028" | "\u2029" | "\r\n"

    comment = multiLineComment | singleLineComment

    multiLineComment = "/*" (~"*/" sourceCharacter)* "*/"
    singleLineComment = "//" (~lineTerminator sourceCharacter)*

}
