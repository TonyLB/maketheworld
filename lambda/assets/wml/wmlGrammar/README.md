The wml.ohm-bundle files are generated:  Only the wml.ohm and wmlQuery.ohm files are original.

To regenerate (e.g. after changing the wml.ohm file), navigate to the assets/wml directory, and
execute this command:  npx ohm generateBundles 'wmlGrammar/*.ohm'

... and then rename the 'ohm-bundle.js' files to 'ohm-bundle.cjs', since the --esm option for
CLI is not released yet, so we can only create CommonJS files, which Lambda needs to be flagged
for.