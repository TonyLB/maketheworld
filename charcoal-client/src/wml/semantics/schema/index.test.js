import wmlGrammar from '../../wmlGrammar/wml.ohm-bundle.js'
import { wmlSemantics } from '../../index.js'

describe('WML semantic schema', () => {
    it('should parse elements properly', () => {
        const match = wmlGrammar.match(`
            <Asset key=(Test) fileName="test">
                <Import from=(BASE)>
                    <Use key=(basePower) as=(power) />
                    <Use key=(overview) />
                </Import>
                <Room key=(ABC)>
                    <Name>Vortex</Name>
                    Vortex
                    <Link key=(switchToggle) to=(toggleOpen)>(toggle)</Link>
                    <Exit from=(DEF)>vortex</Exit>
                </Room>
                <Condition if={open}>
                    <Depend on=(open) />
                    <Room key=(ABC)>
                        <Exit to=(DEF)>welcome</Exit>
                    </Room>
                </Condition>
                <Room key=(DEF)>
                    <Name>Welcome</Name>
                </Room>
                <Variable key=(open) default={false} />
                <Action key=(toggleOpen) src={open = !open} />
                <Computed key=(closed) src={!open}>
                    <Depend on=(open) />
                </Computed>
            </Asset>
        `)
        const schema = wmlSemantics(match).schema()
        expect(schema).toEqual({
            key: 'Test',
            tag: 'Asset',
            fileName: 'test',
            props: {},
            contents: [{
                tag: 'Import',
                from: 'BASE',
                mapping: {
                    power: 'basePower',
                    overview: 'overview'
                },
                contents: [],
                props: {}
            },
            {
                key: 'ABC',
                tag: 'Room',
                name: 'Vortex',
                global: false,
                render: [
                    'Vortex ',
                    {
                        tag: 'Link',
                        key: 'switchToggle',
                        text: '(toggle)',
                        to: 'toggleOpen',
                        props: {},
                        contents: []
                    }
                ],
                props: {},
                contents: [{
                    tag: 'Exit',
                    from: 'DEF',
                    name: 'vortex',
                    props: {},
                    contents: []
                }]
            },
            {
                tag: 'Condition',
                if: 'open',
                dependencies: ['open'],
                props: {},
                contents: [{
                    key: 'ABC',
                    tag: 'Room',
                    global: false,
                    render: [],
                    props: {},
                    contents: [{
                        tag: 'Exit',
                        to: 'DEF',
                        name: 'welcome',
                        props: {},
                        contents: []
                    }]
                }]
            },
            {
                key: 'DEF',
                tag: 'Room',
                name: 'Welcome',
                global: false,
                props: {},
                render: [],
                contents: []
            },
            {
                key: 'open',
                tag: 'Variable',
                default: 'false',
                props: {}
            },
            {
                key: 'toggleOpen',
                tag: 'Action',
                src: 'open = !open',
                props: {}
            },
            {
                key: 'closed',
                tag: 'Computed',
                dependencies: ['open'],
                src: '!open',
                props: {}
            }]
        })
    })

    it('should parse a character element', () => {
        const match = wmlGrammar.match(`
            <Character key=(Tess) fileName="Tess" player="testy">
                <Image key=(icon) fileURL="testIcon.png" />
                <Name>Tess</Name>
                <Pronouns
                    subject="she"
                    object="her"
                    possessive="her"
                    adjective="hers"
                    reflexive="herself"
                />
            </Character>
        `)
        const schema = wmlSemantics(match).schema()
        expect(schema).toEqual({
            key: 'Tess',
            tag: 'Character',
            fileName: 'Tess',
            player: "testy",
            Name: 'Tess',
            Pronouns: {
                subject: 'she',
                object: 'her',
                possessive: 'her',
                adjective: 'hers',
                reflexive: 'herself'
            },
            fileURL: 'testIcon.png',
            props: {},
            contents: []
        })
    })

    it('should parse a story element', () => {
        const match = wmlGrammar.match(`
            <Story key=(Test) instance fileName="test">
                <Room key=(ABC)>
                    <Name>Vortex</Name>
                    Vortex
                </Room>
            </Story>
        `)
        const schema = wmlSemantics(match).schema()
        expect(schema).toEqual({
            key: 'Test',
            tag: 'Asset',
            fileName: 'test',
            Story: true,
            instance: true,
            props: {},
            contents: [{
                key: 'ABC',
                tag: 'Room',
                name: 'Vortex',
                global: false,
                render: [
                    'Vortex '
                ],
                props: {},
                contents: []
            }]
        })
    })

    it('should parse feature elements', () => {
        const match = wmlGrammar.match(`
            <Story key=(Test) instance fileName="test">
                <Feature key=(clockTower)>
                    A clock-tower of weathered grey stone looms over the area.
                </Feature>
                <Room key=(ABC)>
                    <Name>Vortex</Name>
                    <Feature key=(clockTower) />
                    Vortex
                </Room>
            </Story>
        `)
        const schema = wmlSemantics(match).schema()
        expect(schema).toEqual({
            key: 'Test',
            tag: 'Asset',
            fileName: 'test',
            Story: true,
            instance: true,
            props: {},
            contents: [{
                key: 'clockTower',
                tag: 'Feature',
                global: false,
                props: {},
                contents: [],
                render: ['A clock-tower of weathered grey stone looms over the area. ']
            },
            {
                key: 'ABC',
                tag: 'Room',
                name: 'Vortex',
                global: false,
                render: [
                    'Vortex '
                ],
                props: {},
                contents: [{
                    key: 'clockTower',
                    tag: 'Feature',
                    global: false,
                    props: {},
                    contents: [],
                    render: []
                }]
            }]
        })
    })

    it('should parse map elements', () => {
        const match = wmlGrammar.match(`
            <Story key=(Test) instance fileName="test">
                <Map key=(TestMap)>
                    <Name>Test Map</Name>
                    <Image key=(ImageTest) fileURL="https://test.com/imageTest.png" />
                    <Room key=(ABC) x="200" y="150" />
                </Map>
            </Story>
        `)
        const schema = wmlSemantics(match).schema()
        expect(schema).toEqual({
            key: 'Test',
            tag: 'Asset',
            fileName: 'test',
            Story: true,
            instance: true,
            props: {},
            contents: [{
                key: 'TestMap',
                tag: 'Map',
                name: 'Test Map',
                props: {},
                rooms: {
                    ABC: {
                        x: 200,
                        y: 150
                    }
                },
                contents: [{
                    key: 'ImageTest',
                    tag: 'Image',
                    fileURL: 'https://test.com/imageTest.png',
                    props: {},
                    contents: []
                },
                {
                    key: 'ABC',
                    tag: 'Room',
                    global: false,
                    render: [],
                    props: {},
                    contents: []
                }]
            }]
        })
    })
})