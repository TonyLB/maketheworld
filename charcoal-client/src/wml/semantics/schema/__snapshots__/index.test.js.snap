// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WML semantic schema should derive spaceBefore and spaceAfter properties from whitespace in description 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "stone",
      "props": Object {},
      "render": Array [],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Feature",
    },
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "clockTower",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "A clock-tower of weathered grey-",
        },
        Object {
          "contents": Array [],
          "props": Object {},
          "spaceAfter": true,
          "spaceBefore": false,
          "tag": "Link",
          "text": "stone",
          "to": "stone",
        },
        Object {
          "spaceAfter": false,
          "spaceBefore": true,
          "tag": "String",
          "value": "looms over the area.",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Feature",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "display": undefined,
          "global": false,
          "key": "clockTower",
          "props": Object {},
          "render": Array [],
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "Feature",
        },
      ],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": true,
          "spaceBefore": false,
          "tag": "String",
          "value": "A spinning tumble of wreckage, surrounding a",
        },
        Object {
          "contents": Array [],
          "props": Object {},
          "spaceAfter": false,
          "spaceBefore": true,
          "tag": "Link",
          "text": "clock tower",
          "to": "clockTower",
        },
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": ".",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Room",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse a character element 1`] = `
Object {
  "Name": "Tess",
  "Pronouns": Object {
    "adjective": "hers",
    "object": "her",
    "possessive": "her",
    "reflexive": "herself",
    "subject": "she",
  },
  "contents": Array [],
  "fileName": "Tess",
  "fileURL": "testIcon.png",
  "key": "Tess",
  "player": "testy",
  "props": Object {},
  "tag": "Character",
  "zone": "Library",
}
`;

exports[`WML semantic schema should parse a story element 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "name": "Vortex",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "Vortex",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Room",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse elements properly 1`] = `
Object {
  "contents": Array [
    Object {
      "contents": Array [],
      "from": "BASE",
      "mapping": Object {
        "overview": Object {
          "key": "overview",
          "type": "Room",
        },
        "power": Object {
          "key": "basePower",
          "type": "Variable",
        },
      },
      "props": Object {},
      "tag": "Import",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "from": "DEF",
          "key": undefined,
          "name": "vortex",
          "props": Object {},
          "tag": "Exit",
          "to": undefined,
        },
      ],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "name": "Vortex",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": true,
          "spaceBefore": false,
          "tag": "String",
          "value": "Vortex",
        },
        Object {
          "contents": Array [],
          "props": Object {},
          "spaceAfter": false,
          "spaceBefore": true,
          "tag": "Link",
          "text": "(toggle)",
          "to": "toggleOpen",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Room",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [
            Object {
              "contents": Array [],
              "from": undefined,
              "key": undefined,
              "name": "welcome",
              "props": Object {},
              "tag": "Exit",
              "to": "DEF",
            },
          ],
          "display": undefined,
          "global": false,
          "key": "ABC",
          "props": Object {},
          "render": Array [],
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "Room",
        },
      ],
      "dependencies": Array [
        "open",
      ],
      "if": "open",
      "props": Object {},
      "tag": "Condition",
    },
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "DEF",
      "name": "Welcome",
      "props": Object {},
      "render": Array [],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Room",
    },
    Object {
      "default": "false",
      "key": "open",
      "props": Object {},
      "tag": "Variable",
    },
    Object {
      "key": "toggleOpen",
      "props": Object {},
      "src": "open = !open",
      "tag": "Action",
    },
    Object {
      "dependencies": Array [
        "open",
      ],
      "key": "closed",
      "props": Object {},
      "src": "!open",
      "tag": "Computed",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse feature elements 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "clockTower",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "A clock-tower of weathered grey stone looms over the area.",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Feature",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "display": undefined,
          "global": false,
          "key": "clockTower",
          "props": Object {},
          "render": Array [],
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "Feature",
        },
      ],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "name": "Vortex",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "Vortex",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Room",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse map elements 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "fileURL": "https://test.com/imageTest.png",
          "key": "ImageTest",
          "props": Object {},
          "tag": "Image",
        },
        Object {
          "contents": Array [],
          "display": undefined,
          "global": false,
          "key": "ABC",
          "props": Object {},
          "render": Array [],
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "Room",
        },
      ],
      "images": Array [
        "ImageTest",
      ],
      "key": "TestMap",
      "name": "Test Map",
      "props": Object {},
      "rooms": Object {
        "ABC": Object {
          "x": 200,
          "y": 150,
        },
      },
      "tag": "Map",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse multiple and nested conditionals 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "default": "true",
      "key": "weatheredClockTower",
      "props": Object {},
      "tag": "Variable",
    },
    Object {
      "default": "true",
      "key": "brightClockTower",
      "props": Object {},
      "tag": "Variable",
    },
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "clockTower",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "A clock-tower of",
        },
      ],
      "spaceAfter": true,
      "spaceBefore": false,
      "tag": "Feature",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "display": undefined,
          "global": false,
          "key": "clockTower",
          "props": Object {},
          "render": Array [
            Object {
              "spaceAfter": false,
              "spaceBefore": false,
              "tag": "String",
              "value": "weathered",
            },
          ],
          "spaceAfter": true,
          "spaceBefore": true,
          "tag": "Feature",
        },
      ],
      "dependencies": Array [],
      "if": "weatheredClockTower",
      "props": Object {},
      "tag": "Condition",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [
            Object {
              "contents": Array [],
              "display": undefined,
              "global": false,
              "key": "clockTower",
              "props": Object {},
              "render": Array [
                Object {
                  "spaceAfter": false,
                  "spaceBefore": false,
                  "tag": "String",
                  "value": "bright",
                },
              ],
              "spaceAfter": true,
              "spaceBefore": true,
              "tag": "Feature",
            },
          ],
          "dependencies": Array [],
          "if": "!brightClockTower",
          "props": Object {},
          "tag": "Condition",
        },
      ],
      "dependencies": Array [],
      "if": "!weatheredClockTower",
      "props": Object {},
      "tag": "Condition",
    },
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "clockTower",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "grey stone looms over the area.",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": true,
      "tag": "Feature",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should parse multiple unconditioned descriptions on same room 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "test",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "One",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Room",
    },
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "test",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "Two",
        },
      ],
      "spaceAfter": false,
      "spaceBefore": false,
      "tag": "Room",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;

exports[`WML semantic schema should percolate description spaceBefore and spaceAfter up hierarchy 1`] = `
Object {
  "Story": true,
  "contents": Array [
    Object {
      "contents": Array [],
      "display": undefined,
      "global": false,
      "key": "clockTower",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "A clock-tower of weathered grey stone looms over the area.",
        },
      ],
      "spaceAfter": true,
      "spaceBefore": true,
      "tag": "Feature",
    },
    Object {
      "contents": Array [
        Object {
          "contents": Array [],
          "display": undefined,
          "global": false,
          "key": "clockTower",
          "props": Object {},
          "render": Array [],
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "Feature",
        },
      ],
      "display": undefined,
      "global": false,
      "key": "ABC",
      "name": "Vortex",
      "props": Object {},
      "render": Array [
        Object {
          "spaceAfter": false,
          "spaceBefore": false,
          "tag": "String",
          "value": "Vortex",
        },
      ],
      "spaceAfter": true,
      "spaceBefore": true,
      "tag": "Room",
    },
  ],
  "description": undefined,
  "fileName": "test",
  "instance": true,
  "key": "Test",
  "player": undefined,
  "props": Object {},
  "tag": "Asset",
  "zone": undefined,
}
`;
